File: backend/src/main/java/com/safebite/backend/model/Restaurante.java
package com.safebite.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "restaurantes")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Restaurante {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String nombre;

    @Column(length = 2000)
    private String descripcion;

    private String direccion;

    private Double latitud;

    private Double longitud;

    private String imagenUrl;

    @Builder.Default
    private Double ratingPromedio = 0.0;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "restaurante_tipos_cocina", joinColumns = @JoinColumn(name = "restaurante_id"))
    @Column(name = "tipo_cocina")
    @Builder.Default
    private Set<String> tiposCocina = new HashSet<>();

    @ElementCollection(targetClass = TipoIntolerancia.class, fetch = FetchType.EAGER)
    @CollectionTable(name = "restaurante_opciones_aptas", joinColumns = @JoinColumn(name = "restaurante_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "intolerancia")
    @Builder.Default
    private Set<TipoIntolerancia> opcionesAptasPara = new HashSet<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creado_por")
    private Usuario creadoPor;

    @Builder.Default
    private boolean activo = true;

    @Builder.Default
    private LocalDateTime fechaCreacion = LocalDateTime.now();
}

File: frontend/lib/core/models/restaurante.dart
import 'usuario.dart';

class Restaurante {
  final int id;
  final String nombre;
  final String? descripcion;
  final String? direccion;
  final double? latitud;
  final double? longitud;
  final String? imagenUrl;
  final double ratingPromedio;
  final Set<String> tiposCocina;
  final Set<TipoIntolerancia> opcionesAptasPara;

  Restaurante({
    required this.id,
    required this.nombre,
    this.descripcion,
    this.direccion,
    this.latitud,
    this.longitud,
    this.imagenUrl,
    required this.ratingPromedio,
    required this.tiposCocina,
    required this.opcionesAptasPara,
  });

  factory Restaurante.fromJson(Map<String, dynamic> json) {
    return Restaurante(
      id: json['id'],
      nombre: json['nombre'] ?? '',
      descripcion: json['descripcion'],
      direccion: json['direccion'],
      latitud: (json['latitud'] as num?)?.toDouble(),
      longitud: (json['longitud'] as num?)?.toDouble(),
      imagenUrl: json['imagenUrl'],
      ratingPromedio: (json['ratingPromedio'] as num?)?.toDouble() ?? 0.0,
      tiposCocina: ((json['tiposCocina'] as List?) ?? []).map((e) => e.toString()).toSet(),
      opcionesAptasPara: ((json['opcionesAptasPara'] as List?) ?? [])
          .map((e) => tipoIntoleranciaDesde(e))
          .toSet(),
    );
  }
}

class Comentario {
  final int id;
  final int restauranteId;
  final String usuarioNombre;
  final String texto;
  final DateTime fecha;

  Comentario({
    required this.id,
    required this.restauranteId,
    required this.usuarioNombre,
    required this.texto,
    required this.fecha,
  });

  factory Comentario.fromJson(Map<String, dynamic> json) => Comentario(
        id: json['id'],
        restauranteId: json['restauranteId'],
        usuarioNombre: json['usuarioNombre'] ?? '',
        texto: json['texto'] ?? '',
        fecha: DateTime.tryParse(json['fecha'] ?? '') ?? DateTime.now(),
      );
}

class Evaluacion {
  final int id;
  final int restauranteId;
  final String usuarioNombre;
  final int puntuacion;
  final String? comentario;
  final DateTime fecha;

  Evaluacion({
    required this.id,
    required this.restauranteId,
    required this.usuarioNombre,
    required this.puntuacion,
    this.comentario,
    required this.fecha,
  });

  factory Evaluacion.fromJson(Map<String, dynamic> json) => Evaluacion(
        id: json['id'],
        restauranteId: json['restauranteId'],
        usuarioNombre: json['usuarioNombre'] ?? '',
        puntuacion: json['puntuacion'] ?? 0,
        comentario: json['comentario'],
        fecha: DateTime.tryParse(json['fecha'] ?? '') ?? DateTime.now(),
      );
}
